File: predictor_api/predictor_api/service/tournament_template_service.py
from typing import Any
from uuid import UUID

from fastapi import HTTPException

from db_handler.db_handler.model.column import Column
from db_handler.db_handler.model.query_condition import QueryCondition
from db_handler.db_handler.model.query_condition_group import (
    QueryConditionGroup
)
from db_handler.db_handler.model.query_request import QueryRequest
from db_handler.db_handler.model.query_response import QueryResponse
from db_handler.db_handler.model.table import Table
from db_handler.db_handler.model.table_join import TableJoin
from db_handler.db_handler.model.type.condition_operator import (
    ConditionOperator
)
from db_handler.db_handler.model.type.sql_operator import SqlOperator
from db_handler.db_handler.model.type.table_join_type import TableJoinType
from db_handler.db_handler.model.update_request import UpdateRequest
from db_handler.db_handler.service.database_query_service import (
    DatabaseQueryService
)
from db_handler.db_handler.util.store_constants import StoreConstants
from predictor_api.predictor_api.model.knockout_template import (
    KnockoutTemplate
)
from predictor_api.predictor_api.model.league_template import LeagueTemplate
from predictor_api.predictor_api.model.tournament import Tournament
from predictor_api.predictor_api.model.tournament_template import (
    TournamentTemplate
)
from predictor_api.predictor_api.model.tournament_template_request import (
    TournamentTemplateRequest
)
from predictor_api.predictor_api.service.knockout_template_service import (
    KnockoutTemplateService
)
from predictor_api.predictor_api.util.predictor_constants import (
    PredictorConstants
)


class TournamentTemplateService:
    """
    Service for performing tournament template-related actions.

    Attributes:
        __query_service (DatabaseQueryService): The database query service.
    """

    def __init__(
            self,
            database_query_service: DatabaseQueryService,
            knockout_template_service: KnockoutTemplateService):
        """
        Initialise the TournamentTemplateService.

        Args:
            database_query_service (DatabaseQueryService): The database query
                service.
        """
        self.__query_service = database_query_service
        self.__knockout_template_service = knockout_template_service

    def get_tournament_templates(self) -> list[TournamentTemplate]:
        """
        Retrieve stored tournament templates.

        Returns:
            list[TournamentTemplate]: The stored tournament templates.
        """
        query_request: QueryRequest = QueryRequest(
            columns=[
                Column.of("tourn", StoreConstants.ID),
                Column(
                    parts=["tourn", "name"],
                    alias="tournamentName"
                ),
                Column.of("tourn", "knockoutTemplateId"),
                Column(
                    parts=["league", StoreConstants.ID],
                    alias="leagueId"
                ),
                Column(
                    parts=["league", "name"],
                    alias="leagueName"
                ),
                Column.of("league", "groupCount"),
                Column.of("league", "teamsPerGroup"),
                Column.of("league", "homeAndAway")
            ],
            table=Table.of(
                PredictorConstants.PREDICTOR_SCHEMA,
                TournamentTemplate.TARGET_TABLE,
                "tourn"
            ),
            tableJoins=[
                TableJoin.of(
                    Table.of(
                        PredictorConstants.PREDICTOR_SCHEMA,
                        LeagueTemplate.TARGET_TABLE,
                        "league"
                    ),
                    QueryCondition.of(
                        Column.of("tourn", "leagueTemplateId"),
                        Column.of("league", StoreConstants.ID)
                    ),
                    TableJoinType.LEFT
                )
            ]
        )

        query_response: QueryResponse = (
            self.__query_service.retrieve_records(query_request)
        )

        return list(
            map(
                lambda record:
                self.__build_tournament_template(record),
                query_response.records
            )
        )

    def create_tournament_templates(
            self,
            tournament_templates: list[TournamentTemplateRequest]
    ) -> list[TournamentTemplate]:
        """
        Create new tournament templates.

        Args:
            tournament_templates (list[TournamentTemplateRequest]): The new
                tournament templates to create.

        Returns:
            list[TournamentTemplate]: The newly created tournament templates.
        """
        records: list[dict[str, Any]] = list(
            map(
                lambda tournament_template:
                tournament_template.model_dump(),
                tournament_templates
            )
        )

        update_request: UpdateRequest = UpdateRequest(
            operation=SqlOperator.INSERT,
            table=Table.of(
                PredictorConstants.PREDICTOR_SCHEMA,
                TournamentTemplate.TARGET_TABLE
            ),
            records=records
        )

        self.__query_service.update_records(update_request)

        included_tournaments: list[UUID] = list(
            map(
                lambda tournament_template:
                tournament_template.id,
                tournament_templates
            )
        )

        query_request: QueryRequest = QueryRequest(
            columns=[
                Column.of("tourn", StoreConstants.ID),
                Column(
                    parts=["tourn", "name"],
                    alias="tournamentName"
                ),
                Column.of("tourn", "knockoutTemplateId"),
                Column(
                    parts=["league", StoreConstants.ID],
                    alias="leagueId"
                ),
                Column(
                    parts=["league", "name"],
                    alias="leagueName"
                ),
                Column.of("league", "groupCount"),
                Column.of("league", "teamsPerGroup"),
                Column.of("league", "homeAndAway")
            ],
            table=Table.of(
                PredictorConstants.PREDICTOR_SCHEMA,
                TournamentTemplate.TARGET_TABLE,
                "tourn"
            ),
            tableJoins=[
                TableJoin.of(
                    Table.of(
                        PredictorConstants.PREDICTOR_SCHEMA,
                        LeagueTemplate.TARGET_TABLE,
                        "league"
                    ),
                    QueryCondition.of(
                        Column.of("tourn", "leagueTemplateId"),
                        Column.of("league", StoreConstants.ID)
                    ),
                    TableJoinType.LEFT
                )
            ],
            conditionGroup=QueryConditionGroup.of(
                QueryCondition(
                    column=Column.of("tourn", StoreConstants.ID),
                    operator=ConditionOperator.IN,
                    value=included_tournaments
                )
            )
        )

        query_response: QueryResponse = (
            self.__query_service.retrieve_records(query_request)
        )

        return list(
            map(
                lambda record:
                self.__build_tournament_template(record),
                query_response.records
            )
        )

    def get_tournament_template_by_id(
            self,
            tournament_template_id: UUID) -> TournamentTemplate:
        """
        Retrieve a single stored tournament template by its id.

        Args:
            tournament_template_id (UUID): The id of the tournament template
                to retrieve.

        Returns:
            TournamentTemplate: The retrieved tournament template.
        """
        query_request: QueryRequest = QueryRequest(
            columns=[
                Column.of("tourn", StoreConstants.ID),
                Column(
                    parts=["tourn", "name"],
                    alias="tournamentName"
                ),
                Column.of("tourn", "knockoutTemplateId"),
                Column(
                    parts=["league", StoreConstants.ID],
                    alias="leagueId"
                ),
                Column(
                    parts=["league", "name"],
                    alias="leagueName"
                ),
                Column.of("league", "groupCount"),
                Column.of("league", "teamsPerGroup"),
                Column.of("league", "homeAndAway")
            ],
            table=Table.of(
                PredictorConstants.PREDICTOR_SCHEMA,
                TournamentTemplate.TARGET_TABLE,
                "tourn"
            ),
            tableJoins=[
                TableJoin.of(
                    Table.of(
                        PredictorConstants.PREDICTOR_SCHEMA,
                        LeagueTemplate.TARGET_TABLE,
                        "league"
                    ),
                    QueryCondition.of(
                        Column.of("tourn", "leagueTemplateId"),
                        Column.of("league", StoreConstants.ID)
                    ),
                    TableJoinType.LEFT
                )
            ],
            conditionGroup=QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of("tourn", StoreConstants.ID),
                    tournament_template_id
                )
            )
        )

        query_response: QueryResponse = (
            self.__query_service.retrieve_records(query_request)
        )

        if len(query_response.records) == 0:
            raise HTTPException(
                status_code=404,
                detail="No tournament templates found with a matching id."
            )

        return list(
            map(
                lambda record:
                self.__build_tournament_template(record),
                query_response.records
            )
        )[0]

    def delete_tournament_template_by_id(self, tournament_template_id: UUID):
        """
        Delete a single stored tournament template by its id.

        Args:
            tournament_template_id (UUID): The id of the tournament template
                to delete.
        """
        query_request: QueryRequest = QueryRequest(
            table=Table.of(
                PredictorConstants.PREDICTOR_SCHEMA,
                Tournament.TARGET_TABLE
            ),
            conditionGroup=QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of("templateId"),
                    tournament_template_id
                )
            )
        )

        query_response: QueryResponse = (
            self.__query_service.retrieve_records(query_request)
        )

        if query_response.recordCount > 0:
            raise HTTPException(
                status_code=409,
                detail="Cannot delete tournament template as it is part of an "
                       "existing tournament."
            )

        update_request: UpdateRequest = UpdateRequest(
            operation=SqlOperator.DELETE,
            table=Table.of(
                PredictorConstants.PREDICTOR_SCHEMA,
                TournamentTemplate.TARGET_TABLE
            ),
            conditionGroup=QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of(StoreConstants.ID),
                    tournament_template_id
                )
            )
        )

        self.__query_service.update_records(update_request)

    def __build_tournament_template(
            self,
            record: dict[str, Any]) -> TournamentTemplate:
        """
        Build a TournamentTemplate object with child LeagueTemplate and
        KnockoutTemplate objects from a database record.

        Args:
            record (dict[str, Any]): The database record.

        Returns:
            TournamentTemplate: The newly created TournamentTemplate object.
        """
        template: TournamentTemplate = TournamentTemplate(
            id=record[StoreConstants.ID],
            name=record["tournamentName"]
        )

        if "leagueId" in record and record["leagueId"] is not None:
            template = template.model_copy(update={
                "league": LeagueTemplate(
                    id=record["leagueId"],
                    name=record["leagueName"],
                    groupCount=record["groupCount"],
                    teamsPerGroup=record["teamsPerGroup"],
                    homeAndAway=record["homeAndAway"]
                )
            })

        if ("knockoutTemplateId" in record and
                record["knockoutTemplateId"] is not None):
            knock_template: KnockoutTemplate = (
                self.__knockout_template_service.get_knockout_template_by_id(
                    record["knockoutTemplateId"]
                )
            )

            template = template.model_copy(update={
                "knockout": knock_template
            })

        return template

File: src/main/java/com/kanemullett/service/TournamentTemplateService.java
package com.kanemullett.service;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.kanemullett.model.Column;
import com.kanemullett.model.DatabaseRecord;
import com.kanemullett.model.ImmutableColumn;
import com.kanemullett.model.ImmutableKnockoutTemplate;
import com.kanemullett.model.ImmutableLeagueTemplate;
import com.kanemullett.model.ImmutableQueryCondition;
import com.kanemullett.model.ImmutableQueryRequest;
import com.kanemullett.model.ImmutableTournamentTemplate;
import com.kanemullett.model.ImmutableUpdateRequest;
import com.kanemullett.model.KnockoutTemplate;
import com.kanemullett.model.LeagueTemplate;
import com.kanemullett.model.QueryCondition;
import com.kanemullett.model.QueryConditionGroup;
import com.kanemullett.model.QueryRequest;
import com.kanemullett.model.QueryResponse;
import com.kanemullett.model.RoundTemplate;
import com.kanemullett.model.Table;
import com.kanemullett.model.TableJoin;
import com.kanemullett.model.Tournament;
import com.kanemullett.model.TournamentTemplate;
import com.kanemullett.model.TournamentTemplateRecord;
import com.kanemullett.model.UpdateRequest;
import com.kanemullett.model.type.ConditionOperator;
import com.kanemullett.model.type.JoinType;
import com.kanemullett.model.type.SqlOperator;
import com.kanemullett.util.DatabaseConstants;
import com.kanemullett.util.PredictorConstants;

import jakarta.annotation.Nullable;

@Service
public class TournamentTemplateService {

    private final DatabaseQueryService queryService;

    private final ObjectMapper mapper = new ObjectMapper();

    private static final String TOURNAMENT_TEMPLATES_TABLE_ALIAS = "tt";
    private static final String LEAGUE_TEMPLATES_TABLE_ALIAS = "lt";
    private static final String KNOCKOUT_TEMPLATES_TABLE_ALIAS = "kt";

    private static final String TOURNAMENT_TEMPLATE_NAME_COLUMN = "tournamentTemplateName";
    private static final String LEAGUE_TEMPLATE_NAME_COLUMN = "leagueTemplateName";
    private static final String KNOCKOUT_TEMPLATE_NAME_COLUMN = "knockoutTemplateName";

    public TournamentTemplateService(DatabaseQueryService queryService) {
        this.queryService = queryService;
    }

    public List<TournamentTemplate> getTournamentTemplates() {
        return getTournamentTemplates(null);
    }

    public List<TournamentTemplate> createTournamentTemplates(List<TournamentTemplateRecord> tournamentTemplates) {
        final UpdateRequest<TournamentTemplateRecord> updateRequest = ImmutableUpdateRequest.<TournamentTemplateRecord>builder()
            .operation(SqlOperator.INSERT)
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, TournamentTemplateRecord.TARGET_TABLE))
            .records(tournamentTemplates)
            .build();

        queryService.updateRecords(updateRequest);

        return getTournamentTemplates(QueryConditionGroup.of(
            ImmutableQueryCondition.builder()
                .column(Column.of(DatabaseConstants.ID))
                .operator(ConditionOperator.IN)
                .value(tournamentTemplates.stream()
                    .map(TournamentTemplateRecord::getId)
                    .toList())
                .build()
        ));
    }

    public TournamentTemplate getTournamentTemplateById(String tournamentTemplateId) {
        final QueryConditionGroup conditionGroup = QueryConditionGroup.of(
            QueryCondition.of(
                Column.of(DatabaseConstants.ID),
                tournamentTemplateId
            )
        );

        final List<TournamentTemplate> tournamentTemplates = getTournamentTemplates(conditionGroup);

        if (tournamentTemplates.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "No tournament templates found with a matching id.");
        }

        return tournamentTemplates.get(0);
    }

    public void deleteTournamentTemplateById(String tournamentTemplateId) {
        final QueryRequest<Tournament> queryRequest = ImmutableQueryRequest.<Tournament>builder()
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, Tournament.TARGET_TABLE))
            .conditionGroup(QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of(Tournament.TEMPLATE_ID_COLUMN),
                    tournamentTemplateId
                )
            ))
            .recordClass(Tournament.class)
            .build();
        
        final QueryResponse<Tournament> response = queryService.retrieveRecords(queryRequest);

        if (response.getRecordCount() > 0) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "Cannot delete tournament template as it is part of an existing tournament.");
        }

        final UpdateRequest<TournamentTemplateRecord> updateRequest = ImmutableUpdateRequest.<TournamentTemplateRecord>builder()
            .operation(SqlOperator.DELETE)
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, TournamentTemplateRecord.TARGET_TABLE))
            .conditionGroup(QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of(DatabaseConstants.ID),
                    tournamentTemplateId
                )
            ))
            .build();

        queryService.updateRecords(updateRequest);

    }

    private List<TournamentTemplate> getTournamentTemplates(@Nullable QueryConditionGroup conditionGroup) {
        final ImmutableQueryRequest.Builder<DatabaseRecord> queryRequest = ImmutableQueryRequest.builder()
            .columns(List.of(
                Column.of(TOURNAMENT_TEMPLATES_TABLE_ALIAS, DatabaseConstants.ID),
                ImmutableColumn.builder()
                    .parts(List.of(TOURNAMENT_TEMPLATES_TABLE_ALIAS, TournamentTemplateRecord.NAME_COLUMN))
                    .alias(TOURNAMENT_TEMPLATE_NAME_COLUMN)
                    .build(),
                Column.of(TOURNAMENT_TEMPLATES_TABLE_ALIAS, TournamentTemplateRecord.LEAGUE_TEMPLATE_ID_COLUMN),
                ImmutableColumn.builder()
                    .parts(List.of(LEAGUE_TEMPLATES_TABLE_ALIAS, LeagueTemplate.NAME_COLUMN))
                    .alias(LEAGUE_TEMPLATE_NAME_COLUMN)
                    .build(),
                Column.of(LEAGUE_TEMPLATES_TABLE_ALIAS, LeagueTemplate.GROUP_COUNT_COLUMN),
                Column.of(LEAGUE_TEMPLATES_TABLE_ALIAS, LeagueTemplate.TEAMS_PER_GROUP_COLUMN),
                Column.of(LEAGUE_TEMPLATES_TABLE_ALIAS, LeagueTemplate.HOME_AND_AWAY_COLUMN),
                ImmutableColumn.builder()
                    .parts(List.of(KNOCKOUT_TEMPLATES_TABLE_ALIAS, KnockoutTemplate.NAME_COLUMN))
                    .alias(KNOCKOUT_TEMPLATE_NAME_COLUMN)
                    .build(),
                Column.of(KNOCKOUT_TEMPLATES_TABLE_ALIAS, KnockoutTemplate.ROUNDS_COLUMN)
            ))
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, TournamentTemplateRecord.TARGET_TABLE, TOURNAMENT_TEMPLATES_TABLE_ALIAS))
            .joins(List.of(
                TableJoin.of(
                    Table.of(PredictorConstants.PREDICTOR_SCHEMA, LeagueTemplate.TARGET_TABLE, LEAGUE_TEMPLATES_TABLE_ALIAS),
                    QueryCondition.of(
                        Column.of(TOURNAMENT_TEMPLATES_TABLE_ALIAS, TournamentTemplateRecord.LEAGUE_TEMPLATE_ID_COLUMN),
                        Column.of(LEAGUE_TEMPLATES_TABLE_ALIAS, DatabaseConstants.ID)
                    ),
                    JoinType.LEFT
                ),
                TableJoin.of(
                    Table.of(PredictorConstants.PREDICTOR_SCHEMA, KnockoutTemplate.TARGET_TABLE, KNOCKOUT_TEMPLATES_TABLE_ALIAS), 
                    QueryCondition.of(
                        Column.of(TOURNAMENT_TEMPLATES_TABLE_ALIAS, TournamentTemplateRecord.KNOCKOUT_TEMPLATE_ID_COLUMN),
                        Column.of(KNOCKOUT_TEMPLATES_TABLE_ALIAS, DatabaseConstants.ID)
                    ),
                    JoinType.LEFT
                )
            ))
            .recordClass(DatabaseRecord.class);

        if (conditionGroup != null) {
            queryRequest.conditionGroup(conditionGroup);
        }

        final QueryResponse<DatabaseRecord> queryResponse = queryService.retrieveRecords(queryRequest.build());

        return queryResponse.getRecords().stream()
            .map(this::buildTournamentTemplate)
            .toList();
    }

    private TournamentTemplate buildTournamentTemplate(DatabaseRecord databaseRecord) {
        final ImmutableTournamentTemplate.Builder builder = ImmutableTournamentTemplate.builder()
            .id(databaseRecord.getId())
            .name((String) databaseRecord.getData().get(TOURNAMENT_TEMPLATE_NAME_COLUMN));

        final String leagueId = (String) databaseRecord.getData().get(TournamentTemplateRecord.LEAGUE_TEMPLATE_ID_COLUMN);
        if (leagueId != null) {
            builder.league(ImmutableLeagueTemplate.builder()
                .id(leagueId)
                .name((String) databaseRecord.getData().get(LEAGUE_TEMPLATE_NAME_COLUMN))
                .groupCount((Integer) databaseRecord.getData().get(LeagueTemplate.GROUP_COUNT_COLUMN))
                .teamsPerGroup((Integer) databaseRecord.getData().get(LeagueTemplate.TEAMS_PER_GROUP_COLUMN))
                .homeAndAway((Boolean) databaseRecord.getData().get(LeagueTemplate.HOME_AND_AWAY_COLUMN))
                .build());
        }

        final String knockoutId = (String) databaseRecord.getData().get(TournamentTemplateRecord.KNOCKOUT_TEMPLATE_ID_COLUMN);
        if (knockoutId != null) {
            builder.knockout(ImmutableKnockoutTemplate.builder()
                .id(knockoutId)
                .name((String) databaseRecord.getData().get(KNOCKOUT_TEMPLATE_NAME_COLUMN))
                .rounds(parseRounds(databaseRecord.getData().get(KnockoutTemplate.ROUNDS_COLUMN)))
                .build());
        }

        return builder.build();
    }

    private List<RoundTemplate> parseRounds(Object rawValue) {
        if (rawValue == null) {
            return null;
        }
        try {
            return mapper.readValue(rawValue.toString(), new TypeReference<List<RoundTemplate>>() {});
        } catch (JsonProcessingException e) {
            throw new IllegalStateException("Unable to parse rounds JSON: " + rawValue, e);
        }
    }
}
